### PR TITLE
🔥 Update: add timeout context middleware

### PR DESCRIPTION
**Pull request by hakankutluay**
## Description

Timeout Context middleware for [Fiber](https://github.com/gofiber/fiber). As a fiber.Handler wrapper, it creates a context with context.WithTimeout and pass it in UserContext.

If the context passed executions (eg. DB ops, Http calls) takes longer than the given duration to return, the timeout error is set and forwarded to the centralized [ErrorHandler](https://docs.gofiber.io/error-handling).

It has no race conditions, ready to use on production.

Fixes https://github.com/gofiber/fiber/issues/2088

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Checklist:

- [x] For new functionalities I follow the inspiration of the express js framework and built them similar in usage
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation - https://github.com/gofiber/docs for https://docs.gofiber.io/
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] If new dependencies exist, I have checked that they are really necessary and agreed with the maintainers/community (we want to have as few dependencies as possible)
- [x] I tried to make my code as fast as possible with as few allocations as possible
- [x] For new code I have written benchmarks so that they can be analyzed and improved

## Commit formatting:

Use emojis on commit messages so it provides an easy way of identifying the purpose or intention of a commit. Check out the emoji cheatsheet here: https://gitmoji.carloscuesta.me/

### CODE REVIEW

**Comment by ReneWerner87** [comment]
hi thanks for the work,

before you go down the checklist and tick off what is already done, i would like to make one comment

can we include this functionality under the current timeout middleware without causing a breaking change ? 
i guess some consumers are confused and later don't know which of the two to use or wonder why 2 exist

if a merge of both doesn't work, what do you think about creating another entry function?

like for example 
```go
timeout.NewWithContext
```
this we can then customize later in v3 and make it controllable via configurations which are injected by changing the strategy, like with the limiter middleware

what do you think ?

**Comment by hakankutluay** [comment]
Hi @ReneWerner87 ,

Thanks for your valuable comments. I got your point.  

This feature cannot be included in the current timeout middleware, current one handles timeout in a different way and has race conditions also not using context.

Having a new entrypoint that you mentioned is much more clear, I've changed implementation in that way. Thanks for your suggestion.

We may also think to replace the current timeout middleware with this one because it has race conditions and satisfies timeout needs.

**Comment by ReneWerner87** [comment]
thx,
don´t forget this file [middleware/timeoutcontext/README.md](https://github.com/gofiber/fiber/pull/2090/files#diff-4735d62e4cd7207b341cd8e931a1c27dc8e138ccfa710d185e3ab7ac52f41ec5)

**Comment by ReneWerner87** [comment]
since the interface is the same or just extended, I would also support the replacement unless you see any other breaking changes ?

**Comment by hakankutluay** [comment]
There is no breaking change on the signature so I replaced the current timeout middleware with the new one.

**Comment by ReneWerner87** [comment]
ok thx

**Comment by vigo** [comment]
great work @hakankutluay thank you!

**Review by efectn — APPROVED** [review]

**Comment by efectn** [comment]
Please create a PR for docs repo

**Comment by hakankutluay** [comment]
[PR](https://github.com/gofiber/docs/pull/280) created on docs repo

**Review by ReneWerner87 — APPROVED** [review]

**Comment by ReneWerner87** [comment]
@hakankutluay have you tested the examples ? where does the parameter come from, don't see it in the route or routes

our consumers are sometimes lazy and copy directly the examples and see how the whole thing behaves

File: middleware/timeout/README.md [hunk 1]
```diff
@@ -50,19 +55,29 @@ func sleepWithContext(ctx context.Context, d time.Duration) error {
 }
 ```
 
-Use with custom error:
+Test it with curl:
+```bash
+curl --location --request GET 'http://localhost:3000/foo?sleepTime=1000' // http 200 (OK)
```
[the remark this answers is not in the document]
Comment (efectn) [reply middleware/timeout/README.md]: Didnt see last commit 😅😅

**Review by efectn — APPROVED** [review]
Update docs pls

File: middleware/timeout/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (hakankutluay) [reply middleware/timeout/README.md]: > Update docs pls

docs are updated

**Comment by hakankutluay** [comment]
docs are also updated

**Pull request merged**